rasterio/_io.pyx

```diff
@@ -2373,9 +2370,11 @@ cdef class BufferedDatasetWriterBase(DatasetWriterBase):
 
         _ = self.meta
         self._env = ExitStack()
-        self._closed = False
 
     def stop(self):
+        if self._hds == NULL:
```

Comment:
This is likely the reason for the error as it didn't protect against being called a second time.